designs/helm-charts/local-bind-volumes.md

```diff
@@ -0,0 +1,119 @@
+# Title: OpenEBS Enhancement Proposal for Mayastor Local Feature
+
+### OEP Number: <To be assigned>
+### Title: Reintroduction of Local Feature in Mayastor for Single-Replica Volumes
+### Authors: @todeb
+### Status: Proposed
+### Created: 2024-11-07
+### Last Updated: 2024-11-07
+### Target Release: ?
+
+---
+
+## Table of Contents
+
+1. [Overview](#overview)
+2. [Motivation](#motivation)
+3. [Goals](#goals)
+4. [Non-Goals](#non-goals)
+5. [Proposal](#proposal)
+6. [User Stories](#user-stories)
+7. [Implementation Details](#implementation-details)
+8. [Drawbacks](#drawbacks)
+9. [Alternatives](#alternatives)
+10. [Unresolved Questions](#unresolved-questions)
+
+---
+
+## 1. Overview
+
+This proposal reintroduces a "local" feature for the Mayastor storage engine in OpenEBS, designed specifically for single-replica volumes. This feature ensures that:
+
+1. A Mayastor Persistent Volume (PV) is created on the same node where the consuming pod is initially scheduled (using `volumeBindingMode: WaitForFirstConsumer`).
+2. Pods are prevented from rescheduling onto nodes where their Mayastor PVs are not present, maintaining strict local attachment to the original node.
+3. Users can configure the local behavior using a `local` flag, which can be set or unset via `kubectl-mayastor` commands.
+
+By implementing these features, the local configuration improves performance and resilience in environments without dedicated storage nodes.
+
+## 2. Motivation
+
+In clusters that do not have dedicated storage nodes, keeping application pods on the same node as their Mayastor volumes is essential. While `volumeBindingMode: WaitForFirstConsumer` is intended to ensure that volumes are created on the same node as their consuming pods, it does not always work as expected with Mayastor volumes. Reintroducing the `local` feature with this proposal enforces both local provisioning and node-level affinity for single-replica volumes, thereby avoiding cross-node data access and potential latency or availability issues.
+
+## 3. Goals
+
+- Implement the local feature for single-replica volumes in Mayastor, ensuring that PVs are created on the same node as their consuming pods when `volumeBindingMode: WaitForFirstConsumer` is used.
+- Prevent pods from being rescheduled to nodes where the associated volume does not exist.
+- Allow users to set or unset the `local` flag for a volume through `kubectl-mayastor`, providing flexibility and administrative control.
```

Comment:
What if you set `local` on an existing volume, residing on node_x and existing pod residing on node_y?
What would be the expected behaviour? The pod consuming the volume would be moved only after existing pod is deleted? (assuming it's some kind of replica set and not a plain Pod without a controller)

Reply:
Do nothing, is up to user to rollout restart or delete pod and kubernetes scheduler will reshedule it then with affinity to local volume.

Reply:
How do we set PV affinity to existing volumes? The affinity is immutable in the PV. Also for a local:true volume, even if we set local to false, once PV affinity is set the application cannot be rescheduled to a new node. So how does the unset flow work?

Reply:
right that might be the issue, possible options I see:
- get rid of the unset, once volume is created cannot be changed.
- delete pv definition retaining data and recreate the new pv
- make a volume replica with openebs.io/csi-node in [mayastor] affinity, and delete original volume

Reply:
unset is not mandatory to be implemented for this, but could be nice to have add on.
3rd point right, it doesn't make any sense as the pv definition has the nodeaffinity and not the mayastor itself, so making more replicas of volume won't change anything in that case.

Reply:
> 3rd point right, it doesn't make any sense 

ok...
Not sure where "making more replicas" is coming from?
My point here is that with a custom scheduler you don't need to use the CSI topology feature, at all. 
At the cost of having to maintain a custom scheduler, which we might want to avoid..

Reply:
oh I see, your 3rd point, right yes I think replicas won't help here.
Anyways my suggestion is to keep this simple, and remove the ability to modify the local for now, that perhaps can be tackled on separate oep?